entry/csp.js

```diff
@@ -0,0 +1,95 @@
+import { createHash } from 'crypto';
+import { parse } from 'node-html-parser';
+import { URL } from 'url';
+
+const scriptTypeIgnoreList = ['application/json', 'application/ld+json'];
+
+const defaultBaseName = 'http://relative-url';
+
+const hashScriptContents = (scriptContents) =>
+  createHash('sha256').update(scriptContents).digest('base64');
+
+export default function createCSPHandler({ extraHosts = [] } = {}) {
+  const hosts = new Set();
+  const shas = new Set();
+
+  const addScriptContents = (contents) => {
+    shas.add(hashScriptContents(contents));
+  };
+
+  const addScriptUrl = (src) => {
+    const { origin } = new URL(src, defaultBaseName);
+
+    if (origin !== defaultBaseName) {
+      hosts.add(origin);
+    }
+  };
+
+  extraHosts.forEach((host) => addScriptUrl(host));
+
+  const processScriptNode = (scriptNode) => {
+    const src = scriptNode.getAttribute('src');
+
+    if (src) {
+      addScriptUrl(src);
+    } else if (
+      !scriptTypeIgnoreList.includes(scriptNode.getAttribute('type'))
+    ) {
+      addScriptContents(scriptNode.firstChild.rawText);
+    }
+  };
+
+  const registerScript = (script) => {
+    const root = parse(script, { script: true });
+
+    if (!root.valid) {
+      console.error(`Invalid script passed to 'csp.registerScript'`);
+    }
+
+    root.querySelectorAll('script').forEach(processScriptNode);
+  };
+
+  const createCSPTag = () => {
+    const policies = [];
+
+    const inlineCspShas = [];
+
+    for (const sha of shas.values()) {
+      inlineCspShas.push(`'sha256-${sha}'`);
+    }
+
+    const scriptSrcPolicy = [
+      `'self'`,
+      ...hosts.values(),
+      ...inlineCspShas,
+    ].join(' ');
+
+    policies.push(`script-src ${scriptSrcPolicy};`);
+
+    return [
+      `<meta http-equiv="Content-Security-Policy" content="`,
+      policies.join(' '),
+      '">',
+    ].join('');
+  };
+
+  const handleHtml = (html) => {
+    const root = parse(html, { script: true });
+
+    if (!root.valid) {
+      throw new Error('Invalid HTML');
```

Comment:
How leniant is `node-html-parser`? It seems like this would be the first time we are validating a consumer's end HTML, which means that this could pick up a few typos in some project's big template strings.
If so, is it worth making this error message a little more verbose?
```suggestion
      throw new Error(`Unable to parse HTML in order to create CSP tag. Check the output of renderDocument for invalid HTML. Error: ${root.error}`);
```